aptos-move/move-examples/token_objects/ambassador/move/sources/ambassador.move

```diff
@@ -0,0 +1,277 @@
+        property_mutator_ref: property_map::MutatorRef,
+        /// Used to emit LevelUpdateEvent
+        level_update_events: event::EventHandle<LevelUpdateEvent>,
+    }
+
+    /// The ambassador level
+    struct AmbassadorLevel has key {
+        ambassador_level: u64,
+    }
+
+    struct LevelUpdateEvent has drop, store {
+        token_address: address,
+        old_level: u64,
+        new_level: u64,
+    }
+
+    /// Initializes the module, creating the ambassador collection
+    fun init_module(sender: &signer) {
+        create_ambassador_collection(sender);
+    }
+
+    /// Creates the ambassador collection
+    fun create_ambassador_collection(creator: &signer) {
+        let description = string::utf8(COLLECTION_DESCRIPTION);
+        let name = string::utf8(COLLECTION_NAME);
+        let uri = string::utf8(COLLECTION_URI);
+
+        let constructor_ref = collection::create_unlimited_collection(
+            creator,
+            description,
+            name,
+            option::none(),
+            uri,
+        );
+        let object_signer = object::generate_signer(&constructor_ref);
+        let ambassador_collection = AmbassadorCollection {
+        };
+        move_to(&object_signer, ambassador_collection);
+    }
+
+    /// Mints an ambassador token
+    public entry fun mint_ambassador_token(
+        creator: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        soul_bound_to: address,
+    ) {
+        let collection = string::utf8(COLLECTION_NAME);
+        // Creates the ambassador token
+        let constructor_ref = token::create_named_token(
+            creator,
+            collection,
+            description,
+            name,
+            option::none(),
+            uri,
+        );
+
+        // Generates the refs.
+        let object_signer = object::generate_signer(&constructor_ref);
+        let transfer_ref = object::generate_transfer_ref(&constructor_ref);
+        let burn_ref = token::generate_burn_ref(&constructor_ref);
+        let property_mutator_ref = property_map::generate_mutator_ref(&constructor_ref);
+        let linear_transfer_ref = object::generate_linear_transfer_ref(&transfer_ref);
+
+        // Transfers the token to the `soul_bound_to` address
+        object::transfer_with_ref(linear_transfer_ref, soul_bound_to);
+
+        // Disables ungated transfer, thus making the token soulbound
+        object::disable_ungated_transfer(&transfer_ref);
+
+        // Initialize the ambassador token
+        let ambassador_token = AmbassadorToken {
+            burn_ref,
+            property_mutator_ref,
+            level_update_events: object::new_event_handle(&object_signer),
+        };
+        move_to(&object_signer, ambassador_token);
+
+        // Initializes the ambassador level
+        move_to(&object_signer, AmbassadorLevel { ambassador_level: 0 });
+
+        // Initialize the property map with the rank property set to 1
```

Comment:
Why not call set_level and set_rank (may be a common internal version that doesn't emit the events)?

Reply:
For both cases such as level and rank, the nature of "initialization" (e.g., `move_to`, `property_map::init`) seems to be a bit different from "update" (e.g., `set_level`, `set_rank`). We may not want our setters to always check whether they have been already initialized properly before updating the values.

Instead, I a bit simplified the initialization step of the property map. Please let me know how it looks.

Reply:
I see that makes sense.